### ISSUE TITLE
Blockchain, Common: The Merge transition detection

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-monorepo#1412]
I was thinking a bit how we can integrate the transition detection and switch from [here](https://eips.ethereum.org/EIPS/eip-3675#definitions) with the terminal total difficulty and the terminal PoW block into our library set.

My current state of thinking is that we can add the `TERMINAL_TOTAL_DIFFICULTY` difficulty as a parameter to the hardfork file in`Common` as a PoW consensus config paramter (e.g. in [mainnet.json](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/common/src/chains/mainnet.json)). I was first thinking about adding this to the HF file but one thing is that this a) has to available actually before Common switches to the Merge and b) the value of this will be different on a per-chain basis. And at the end this is a PoW parameter (this is the moment where the current PoW usage will end).

The `Blockchain` library can then use this parameter to track the total difficulty value. Once the `pow_block.total_difficulty >= TERMINAL_TOTAL_DIFFICULTY` condition is satisfied for a block added to the chain, the `Blockchain` library emits and event `POW_TERMINAL_TOTAL_DIFFICULTY_REACHED` or the like which can be caught by the client and the client can react accordingly.

(this would mean that we would need to extend the `Blockchain` class as an EventEmitter - I remember from when we switched on this for the `Common` library for a new (and first) `hardforkChanged` event this was more or less side effects free - but just to note)

Does this make sense?

I guess we can actually already start integrating this adding this to blockchain and adding 2-3 tests for this but for now excluding (most) of the client part, maybe giving this a first stub to get a first feeling where this could be integrated and give this some first thought what the client would do next (things like not listening to `NEW_BLOCK_HASHES` any more I guess).

**Comment by holgerd77** [comment]
(please read on site, several post-submit changes)

**Comment by acolytec3** [comment]
I like the general thought of this as the same sort of thing occurred to me after reading your thoughts on #1413.  That said, as far as client handling, do we actually need for `blockchain` to emit an event or can we just add logic to the `protocol` code within the client to disregard messages that match `NEW_BLOCK_HASHES` (maybe [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/bfd53752d8d589a38fc3fb41bb67f5f5acb74f9d/packages/client/lib/net/protocol/boundprotocol.ts#L89-L89) once either difficulty is equal to `TERMINAL_TOTAL_DIFFICULTY` or else even just look at the `common.consensusAlgorithm`?  I guess to my way of thinking, the blockchain's common instance is (or should) already be handling the switch to PoS so we should just be able to check that, right?

**Comment by holgerd77** [comment]
I am a bit unsure about just checking `common.consensusAlgorithm()`, I would still assume this needs some kind of event which can directly acted upon. But we can keep this a bit open for now.

**Comment by gabrocheleau** [comment]
I also think it makes sense to add `TERMINAL_TOTAL_DIFFICULTY` to common, indeed we have to keep in mind that we'll use that param:
- In PoW to know when to "turn PoW off"
- In PoS to know that we've transitioned to PoS

Concerning the event, might it be enough to just listen to the 2 events emitted by the consensus layer? See the two last bullet points here: https://eips.ethereum.org/EIPS/eip-3675#specification :

- `POS_CONSENSUS_VALIDATED` An event occurring when the PoS block is validated with respect to the proof-of-stake consensus rules.
- `POS_FORKCHOICE_UPDATED` An event occurring when the state of the proof-of-stake fork choice is updated.

**Comment by gabrocheleau** [comment]
Post-merge, we could add a `PoS` consensus param `TRANSITION_TOTAL_DIFFICULTY` or something like that, which would correspond to the actual total difficulty of the last PoW block (which is `>= TERMINAL_TOTAL_DIFFICULTY`). We could then refer back to that constant for blockchain `totalDifficulty` in PoS, and use that as the genesis totalDifficulty in a pure PoS context.

**Comment by holgerd77** [comment]
@gabrocheleau yeah, I totally didn't think about the fact that the *real* transition difficulty is deviating from `TERMINAL_TOTAL_DIFFICULTY`. Just some thoughts: I think this would be misplaced in `Common` since *this* is a dynamic parameter, it would not make sense to add this to `Common` and hope for some other library to fill in. Since the transition TD can only be derived from within a blockchain context, I would very much think for now that this should then be handled in the `Blockchain` library.

On first thought I would assume that this will also work. There are actually two entry points on this transition moment:

1) The "live" entry point which is derived directly from the `TERMINAL_TOTAL_DIFFICULTY` we will have as a constant in `Common`. This live entry point is only needed in the context of the `Blockchain` library and the client, to detect this moment and then live-change the rule set while processing the live chain.

2) The second entry point is actually the "post-merge" entry point. This will just be the block number on which the merge finally happened, which will become clear after the merge event. We can then hard-code this block number (as we have done for all the other HFs) into `Common`. This can then be used to e.g. execute blocks - e.g. fetched by RPC post-merge - within the VM (just one use case) where the block number of the block si then known.

One thing we can structurally do to bring these two ways of HF switches together is add a new option `hardforkByTD` whereever we already have a `hardforkByBlockNumber` (I think `Blockchain`, `Block` and `VM`). We can then call into `Common.setHardforkByBlockNumber()` (this is what we already do). If this returns `null` (or similar) we fallback to a new `Common.setHardforkByTD()` method which takes in a TD value and see if this is above the threshold from `Common` and sets the HF by that value.

I guess this should already serve well the two release periods we will have around the Merge time with our libraries: one period where the Merge block is not yet added to Common, and one after the Merge where it is added. I think the above implementation would already give us somewhat broad flexibility on both this transition period as well as the general variety of use cases we will have (calling from a dynamic (HF block not known) client context, calling from a static (HF block known) block context.

With these thoughts in mind I have also some tendency to move over the place where to put TERMINAL_TOTAL_DIFFICULTY in Common, and instead of ordering this in as a PoS config parameter I would rather tend to add this additionally to the HF block number sections in the chain files, e.g. [mainnet.json](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/common/src/chains/mainnet.json):

```json
{
  "name": "merge",
  "block": null,
  "td": 12345678,
  "forkash": null
}
```

At the end there is absolutely no PoS specific about this TD switch. This could theoretically be applied as a transition mechanism to any kind of HF, being it a one switching on the consensus engine or not. And with all written above it generally feels more natural and fitting to me to be placed along the block number here.

**Comment by gabrocheleau** [comment]
Yeah I like your idea of adding a `Common.setHardforkByTD` @holgerd77 and using that as a fallback and adding the `td` in the chain files.
